### PR TITLE
Pre commit ci

### LINKED ISSUE #302: Use rustfmt and setup CI and pre-commit to use it
The pre-commit configuration can be similar to https://github.com/mozilla/rust-code-analysis/blob/4a899635cc7280bdc265cc2429bd164cfe61714c/.pre-commit-config.yaml.

### PR DESCRIPTION
**Pull request by DanielChabrowski**
Closes #302 

Technically also related to #127, though we can also add a separate nightly clippy job that is allowed to fail and includes the benchmarks. To close it you probably want to fix clippy warnings.

I had to disable benchmarks in clippy because of the:
```error[E0554]: #![feature] may not be used on the stable release channel```

Let me know what you think, cheers.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -22,20 +22,21 @@ repos:
         language: system
         types: [file, rust]
         entry: cargo fmt -- --check
+        pass_filenames: false
```
Comment (marco-c) [inline .pre-commit-config.yaml:25]: Why is this needed?

File: .travis.yml [hunk 2]
```diff
@@ -109,6 +109,23 @@ matrix:
...
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - python3-pip
+            - python3-setuptools
+            - python3-wheel
+      script:
+        - rustup component add rustfmt clippy
+        - pip3 install pre-commit
```
Comment (marco-c) [inline .travis.yml:126]: Nit: use `--disable-pip-version-check --quiet --no-cache-dir` parameters to pip

**Comment by marco-c** [comment]
> I had to disable benchmarks in clippy because of the:
```error[E0554]: #![feature] may not be used on the stable release channel```

We could use Nightly clippy

File: .pre-commit-config.yaml (hunk 1 above)
Comment (DanielChabrowski) [reply .pre-commit-config.yaml:25]: Cargo fmt doesn't need passing filenames and `find . -name "*.rs"` includes two somehow unformatted, but also empty, main.rs files in tests.

**Comment by DanielChabrowski** [comment]
> > I had to disable benchmarks in clippy because of the:
> > `error[E0554]: #![feature] may not be used on the stable release channel`
> 
> We could use Nightly clippy exclusively, instead of a Nightly allowed to fail and a stable not allowed to fail.

That's what I actually started with but travis couldn't download nightly fmt. So, since the very beginning it was causing issues. Also, something that "might not work" does not integrate well with pre-commit, it would just be annoying instead of being helpful.

File: .travis.yml (hunk 2 above)
Comment (DanielChabrowski) [reply .travis.yml]: Done.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (marco-c) [reply .pre-commit-config.yaml:25]: Got it!

**Comment by marco-c** [comment]
> That's what I actually started with but travis couldn't download nightly fmt. So, since the very beginning it was causing issues. Also, something that "might not work" does not integrate well with pre-commit, it would just be annoying instead of being helpful.

Fair enough.
